src/storage/txn/scheduler.rs

```diff
@@ -431,15 +431,45 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
         SCHED_STAGE_COUNTER_VEC.get(tag).read_finish.inc();
 
         debug!("read command finished"; "cid" => cid);
-        let tctx = self.inner.dequeue_task_context(cid);
-        if let ProcessResult::NextCommand { cmd } = pr {
+        if let ProcessResult::NextCommands { mut cmds } = pr {
+            // joiner reuse the finished command's context, and responsible for joining the other commands' result
+            let joiner = cmds.pop().unwrap();
+            // the rest commands (joinees) should run on their own and notify the joiner their result by using the channel
+            let joinee_count = cmds.len();
+            let mut tctx = self.inner.dequeue_task_context(cid);
             SCHED_STAGE_COUNTER_VEC.get(tag).next_cmd.inc();
-            self.schedule_command(cmd, tctx.cb.unwrap());
+            let cb = tctx.cb.take();
+            let (tx, rx) = std::sync::mpsc::channel();
+            self.schedule_command(
+                joiner,
+                StorageCallback::Boolean(Box::new(move |pr| {
+                    for _ in 0..joinee_count {
+                        if let Err(_e) = rx.recv() {
+                            return cb.unwrap().execute(ProcessResult::Failed {
+                                err: StorageError::from(ErrorInner::Closed),
```

Comment:
What's the proper Error type here?
Or shall I add a new Error type in ErrorInner?